agent/mapcall/src/main/java/randoop/instrument/CallReplacementTransformer.java

```diff
@@ -178,491 +91,303 @@ public CallReplacementTransformer() {}
+    if (isExcludedClass(loader, fullClassName)) {
+      debug_transform.log("transform: ignoring excluded class %s%n", className);
       return null;
     }
 
-    debug_transform.log(
-        "transforming class %s, loader %s - %s%n", className, loader, loader.getParent());
+    debug_transform.log("%ntransform: ENTER %s%n", className);
 
-    // Parse the bytes of the classfile, die on any errors
+    // Parse the bytes of the classfile
     JavaClass c;
     try {
       ClassParser parser = new ClassParser(new ByteArrayInputStream(classfileBuffer), className);
       c = parser.parse();
     } catch (Exception e) {
-      throw new RuntimeException("Unexpected error", e);
+      debug_transform.log("transform: EXIT parse of %s resulted in error %s%n", className, e);
+      throw new RuntimeException("Unexpected parse exception", e);
     }
 
     try {
       // Get the class information
       ClassGen cg = new ClassGen(c);
-      ifact = new InstructionFactory(cg);
-
-      map_calls(cg, className, loader, current_method_map);
-
-      JavaClass njc = cg.getJavaClass();
-      if (debug) njc.dump("/tmp/ret/" + njc.getClassName() + ".class");
-
-      if (true) {
-        return (cg.getJavaClass().getBytes());
+      if (transformClass(cg)) {
+        JavaClass javaClass = cg.getJavaClass();
+        if (MapCallsAgent.debug) {
+          Path filepath = MapCallsAgent.debugPath.resolve(javaClass.getClassName() + ".class");
+          javaClass.dump(filepath.toFile());
+        }
+        debug_transform.log("transform: EXIT class %s transformed", className);
+        return javaClass.getBytes();
       } else {
-        debug_transform.log("not including class %s (filtered out)", className);
+        debug_transform.log("transform: EXIT class %s not transformed", className);
         return null;
       }
-
     } catch (Throwable e) {
-      out.format("Unexpected error %s in transform", e);
+      debug_transform.log(
+          "transform: EXIT transform of %s resulted in exception %s%n", className, e);
+      System.out.format(
+          "Unexpected exception %s (%s) in class transform of %s%n", e, e.getCause(), className);
       e.printStackTrace();
-      return (null);
+      return null;
     }
   }
 
   /**
-   * Processes each method in cg replacing any specified calls with static user calls.
+   * Indicates whether the class is a non-AWT/Swing boot-loaded class.
+   *
+   * <p>This check is for performance, since attempting to transform all of {@code rt.jar} is
+   * unnecessary. However, we want to transform AWT/Swing classes if they are loaded.
    *
-   * @param fullClassName must be packageName.className
+   * <p>Actually checks whether the class is loaded by the bootstrap loader or by the first
+   * classloader. The first classloader will either be a user-provided boot loader, or the extension
+   * class loader. Since the predicate is mostly for performance, we don't make the extra check to
```

Comment:
If it's "mostly" for performance, what are the other motivations?  Above it says "This check is for performance".